feat: testing discovery and dialing with nwaku integration

gabrielmer opened the pull request:
## Description

Testing DNS discovery and peer connections and disconnections when integrating nwaku.
I commented everything in `api.go`, `api_test.go`, `message_publishing.go` and a big part of `nwaku.go`. I propose to remove the comments incrementally while we implement and test new features with the nwaku integration.



Important changes:
- [x] commenting `api.go`, `api_test.go`, `message_publishing.go` and part of `nwaku.go`
- [x] changed C function definitions to static to avoid compilation error
- [x] added `cGoWakuDialPeerById` and `cGoWakuDisconnectPeerById` using newly added bindings  
- [x] extended `WakuConfig` with more nwaku node configurations 
- [x] changed `cfg` field in `Waku` to be of type `*WakuConfig` 
- [x] added `WakuDialPeerById` and `DisconnectPeerById` functions 
- [x] adapting `TestBasicWakuV2`to use nwkaku's DNS discovery, dialing and disconnecting from peers and checking that it indeed works 

## How to test it
1. compile libwaku in nwaku's `chore-adding-libwaku-dial-and-disconnect-by-id` branch to have the new libwaku's procs that are used
2. In another terminal, run
```
IP_ADDRESS=$(hostname -I | awk '{print $1}');
docker run \
-p 61000:61000/tcp -p 8000:8000/udp -p 8646:8646/tcp harbor.status.im/wakuorg/nwaku:v0.33.0 \
--discv5-discovery=true --cluster-id=16 --log-level=DEBUG \
--nat=extip:${IP_ADDRESS} --discv5-udp-port=8000 --rest-address=0.0.0.0 --store --rest-port=8646 \
```
3. Compile and execute `TestBasicWakuV2` by running
```
go test -tags="use_nwaku,gowaku_no_rln" -run TestBasicWakuV2 ./wakuv2/... -count 1 -v -json | jq -r '.Output'
```

Closes https://github.com/waku-org/nwaku/issues/3039

gabrielmer committed:
fixing setup and adding debug logs

gabrielmer committed:
removing unnecessary channel

gabrielmer committed:
commenting a lot of things and some config changes to start testing

gabrielmer committed:
adding discv5Discovery param

gabrielmer committed:
adding clusterId and shards

gabrielmer committed:
adding dnsdiscovery, staticnodes and some cleanup

gabrielmer committed:
check for connection

gabrielmer committed:
adding disconnect functionality

gabrielmer committed:
adding dialing to test

ghost commented:
Hey @gabrielmer, and thank you so much for making your first pull request in status-go! :heart: Please help us make your experience better by filling out this brief questionnaire https://goo.gl/forms/uWqNcVpVz7OIopXg2

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (30)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_multiplication_x: | 0b5e20b0 | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5940/1/) | 2024-10-14 07:27:40 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5940/1//consoleText) |
| :x: | 0b5e20b0 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5940/1/) | 2024-10-14 07:27:43 | ~1 min | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5940/1/consoleText) |
| :heavy_multiplication_x: | 0b5e20b0 | [#1](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5940/1/) | 2024-10-14 07:27:55 | ~1 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5940/1//consoleText) |
| :x: | 0b5e20b0 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5940/1/) | 2024-10-14 07:28:05 | ~1 min | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5940/1/consoleText) |
| :x: | 0b5e20b0 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5940/1/) | 2024-10-14 07:28:13 | ~1 min | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5940/1/consoleText) |
| | | | | | | |
| :x: | d1f09c1c | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5940/2/) | 2024-10-14 07:38:59 | ~26 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5940/2/consoleText) |
| :x: | d1f09c1c | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5940/2/) | 2024-10-14 07:39:07 | ~31 sec | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5940/2/consoleText) |
| :x: | d1f09c1c | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5940/2/) | 2024-10-14 07:39:19 | ~44 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5940/2/consoleText) |
| :heavy_multiplication_x: | d1f09c1c | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5940/2/) | 2024-10-14 07:39:24 | ~43 sec | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5940/2//consoleText) |
| :heavy_multiplication_x: | d1f09c1c | [#2](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5940/2/) | 2024-10-14 07:39:40 | ~1 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5940/2//consoleText) |
| | | | | | | |
| :x: | 4b9bee46 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5940/3/) | 2024-10-14 08:40:37 | ~25 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5940/3/consoleText) |
| :x: | 4b9bee46 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5940/3/) | 2024-10-14 08:40:41 | ~28 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5940/3/consoleText) |
| :x: | 4b9bee46 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5940/3/) | 2024-10-14 08:40:47 | ~31 sec | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5940/3/consoleText) |
| :heavy_multiplication_x: | 4b9bee46 | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5940/3/) | 2024-10-14 08:41:04 | ~44 sec | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5940/3//consoleText) |
| :heavy_multiplication_x: | 4b9bee46 | [#3](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5940/3/) | 2024-10-14 08:41:18 | ~1 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5940/3//consoleText) |
| | | | | | | |
| :x: | cd41b136 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5940/4/) | 2024-10-14 08:42:44 | ~23 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5940/4/consoleText) |
| :x: | cd41b136 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5940/4/) | 2024-10-14 08:42:45 | ~26 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5940/4/consoleText) |
| :x: | cd41b136 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5940/4/) | 2024-10-14 08:42:52 | ~31 sec | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5940/4/consoleText) |
| :heavy_multiplication_x: | cd41b136 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5940/4/) | 2024-10-14 08:43:10 | ~43 sec | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5940/4//consoleText) |
| :heavy_multiplication_x: | cd41b136 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5940/4/) | 2024-10-14 08:43:25 | ~1 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5940/4//consoleText) |
| | | | | | | |
| :x: | 9373407a | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5940/5/) | 2024-10-14 08:56:20 | ~25 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5940/5/consoleText) |
| :x: | 9373407a | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5940/5/) | 2024-10-14 08:56:21 | ~27 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5940/5/consoleText) |
| :x: | 9373407a | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5940/5/) | 2024-10-14 08:56:28 | ~31 sec | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5940/5/consoleText) |
| :heavy_multiplication_x: | 9373407a | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5940/5/) | 2024-10-14 08:56:54 | ~53 sec | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5940/5//consoleText) |
| :heavy_multiplication_x: | 9373407a | [#5](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5940/5/) | 2024-10-14 08:57:01 | ~1 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5940/5//consoleText) |
| | | | | | | |
| :x: | 524b5661 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5940/6/) | 2024-10-15 11:13:57 | ~28 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5940/6/consoleText) |
| :x: | 524b5661 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5940/6/) | 2024-10-15 11:14:05 | ~33 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5940/6/consoleText) |
| :heavy_multiplication_x: | 524b5661 | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5940/6/) | 2024-10-15 11:14:34 | ~59 sec | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5940/6//consoleText) |
| :heavy_multiplication_x: | 524b5661 | [#6](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5940/6/) | 2024-10-15 11:14:51 | ~1 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5940/6//consoleText) |
| :x: | 524b5661 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5940/6/) | 2024-10-15 11:15:00 | ~1 min | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5940/6/consoleText) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | 37d7fd44 | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5940/7/) | 2024-10-15 11:17:10 | ~27 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5940/7/consoleText) |
| :x: | 37d7fd44 | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5940/7/) | 2024-10-15 11:17:19 | ~31 sec | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5940/7/consoleText) |
| :x: | 37d7fd44 | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5940/7/) | 2024-10-15 11:17:28 | ~44 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5940/7/consoleText) |
| :heavy_multiplication_x: | 37d7fd44 | [#7](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5940/7/) | 2024-10-15 11:17:50 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5940/7//consoleText) |
| :heavy_multiplication_x: | 37d7fd44 | [#7](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5940/7/) | 2024-10-15 11:17:59 | ~1 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5940/7//consoleText) |
| | | | | | | |
| :x: | 5ec90655 | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5940/8/) | 2024-10-15 14:32:38 | ~26 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5940/8/consoleText) |
| :x: | 5ec90655 | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5940/8/) | 2024-10-15 14:32:43 | ~27 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5940/8/consoleText) |
| :x: | 5ec90655 | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5940/8/) | 2024-10-15 14:32:47 | ~31 sec | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5940/8/consoleText) |
| :heavy_multiplication_x: | 5ec90655 | [#8](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5940/8/) | 2024-10-15 14:33:13 | ~53 sec | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5940/8//consoleText) |
| :heavy_multiplication_x: | 5ec90655 | [#8](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5940/8/) | 2024-10-15 14:33:35 | ~1 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5940/8//consoleText) |

gabrielmer committed:
cleanup

gabrielmer committed:
commenting compilation errors

gabrielmer committed:
deleting some tabs

gabrielmer committed:
improving comments

Ivansete-status reviewed (approved):
LGTM! Thanks for it! :100: 
In all the commented code, shall we add the tag "TODO-nwaku", or sth like that, so that we can revisit all of them in the future?

gabrielmer committed:
adding TODO comments

gabrielmer committed:
more TODO comments

gabrielmer commented:
> LGTM! Thanks for it! 💯 In all the commented code, shall we add the tag "TODO-nwaku", or sth like that, so that we can revisit all of them in the future?

Ooh great idea! Just updated it and added "TODO-nwaku" to all the newly commented blocks :)

Thank you!

richard-ramos commented on wakuv2/nwaku_test.go:
```diff
@@ -161,80 +154,87 @@ func parseNodes(rec []string) []*enode.Node {
+		NodeKey:     "11d0dcea28e86f81937a3bd1163473c7fbc0a0db54fd72914849bc47bdf78710",
+		EnableRelay: true,
+		LogLevel:    "DEBUG",
+		DnsDiscoveryUrl: "enrtree://AMOJVZX4V6EXP7NTJPMAYJYST2QP6AJXYW76IU6VGJS7UVSNDYZG4@boot.prod.status.nodes.status.im",
```
This while it will work, it has the inconvenient of polluting the bootnodes with nodes that will only live when we run the tests (hence why the fake enr tree was being created)

gabrielmer replied:
Makes sense 100%! As we discussed, I created the following issue to track this improvement https://github.com/waku-org/nwaku/issues/3119

Thanks so much!

richard-ramos commented on wakuv2/nwaku_test.go:
```diff
@@ -161,80 +154,87 @@ func parseNodes(rec []string) []*enode.Node {
-			return errors.New("no peers discovered")
+		
+		numConnected, err := w.GetNumConnectedPeers()
+		fmt.Println("numConnected: ", numConnected)
```
Might be better to use a log here instead.

gabrielmer replied:
Oops forgot about that! I added it for debugging purposes but it actually can spam a lot, so if it's ok I think I'll just get rid it.

Thank you!

richard-ramos reviewed (approved):
Overall LGTM. Just left a comment regarding the enrtree usage!

gabrielmer committed:
removing spammy log
